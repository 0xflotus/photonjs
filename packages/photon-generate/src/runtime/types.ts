import { DMMF } from './dmmf-types'

export interface ArgError {
  path: string[]
  error: InvalidArgError
}

export interface FieldError {
  path: string[]
  error: InvalidFieldError
}

export type InvalidFieldError = InvalidFieldNameError | InvalidFieldTypeError

export interface InvalidFieldTypeError {
  type: 'invalidFieldType'
  modelName: string
  fieldName: string
  providedValue: any
}

export interface InvalidFieldNameError {
  type: 'invalidFieldName'
  modelName: string
  didYouMean?: string
  providedName: string
}

export type JavaScriptPrimitiveType = 'number' | 'string' | 'boolean'

export type InvalidArgError = InvalidArgNameError | MissingArgError | InvalidArgTypeError

/**
 * This error occurs if the user provides an arg name that doens't exist
 */
export type InvalidArgNameError = {
  type: 'invalidName'
  providedName: string
  providedValue: any
  didYouMeanArg?: string // if the possible names are too different and therefore just arbitrary, we don't suggest anything
  didYouMeanField?: string // if it's very similar to a field, they probably just forgot the select statement
  originalType: string | DMMF.InputType
  outputType?: DMMF.OutputType
}

/**
 * Opposite of InvalidArgNameError - if the user *doesn't* provide an arg that should be provided
 * This error both happens with an implicit and explicit `undefined`
 */
export type MissingArgError = {
  type: 'missingArg'
  missingName: string
  missingType: string | DMMF.InputType // note that this could be an object or scalar type. in the object case, we print the whole object type
  isScalar: boolean // useful for error printing
  isList: boolean
}

/**
 * If the scalar type of an arg is not matching what is required
 */
export type InvalidArgTypeError = {
  type: 'invalidType'
  argName: string
  requiredType: {
    type: string | DMMF.InputType
    isRequired: boolean
    isScalar: boolean
    isList: boolean
  }
  providedValue: any
}
