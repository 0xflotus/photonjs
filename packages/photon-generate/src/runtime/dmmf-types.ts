export namespace DMMF {
  export interface Document {
    datamodel: Datamodel
    schema: Schema
    mappings: Mapping[]
  }

  export interface Enum {
    name: string
    values: string[]
  }

  export interface Datamodel {
    models: Model[]
    enums: Enum[]
  }

  export interface Model {
    name: string
    isEmbedded: boolean
    dbName: string
    fields: Field[]
  }

  export type FieldKind = 'scalar' | 'relation' | 'enum'

  export interface Field {
    kind: FieldKind
    name: string
    isRequired: boolean
    isList: boolean
    isUnique: boolean
    isId: boolean
    type: string
  }

  export interface Schema {
    queries: Query[]
    mutations: Query[]
    inputTypes: InputType[]
    outputTypes: OutputType[]
    enums: Enum[]
  }

  export interface Query {
    name: string
    args: SchemaArg[]
    output: QueryOutput
  }

  export interface QueryOutput {
    name: string
    isRequired: boolean
    isList: boolean
  }

  export interface SchemaArg {
    name: string
    type: string | InputType | Enum
    isScalar: boolean
    isRequired: boolean
    isEnum: boolean
    isList: boolean
  }

  export interface OutputType {
    name: string
    fields: SchemaField[]
  }

  export interface MergedOutputType extends OutputType {
    isEmbedded: boolean
    fields: SchemaField[]
  }

  export interface SchemaField {
    name: string
    type: string | MergedOutputType | Enum // note that in the serialized state we don't have the reference to MergedOutputTypes
    isList: boolean
    isRequired: boolean
    kind: FieldKind
    args: SchemaArg[]
  }

  export interface InputType {
    name: string
    args: SchemaArg[]
  }

  export interface Mapping {
    model: string
    findOne?: string
    findMany?: string
    create?: string
    update?: string
    updateMany?: string
    upsert?: string
    delete?: string
    deleteMany?: string
  }

  export enum ModelAction {
    findOne = 'findOne',
    findMany = 'findMany',
    create = 'create',
    update = 'update',
    updateMany = 'updateMany',
    upsert = 'upsert',
    delete = 'delete',
    deleteMany = 'deleteMany',
  }
}

export interface BaseField {
  name: string
  type: string | DMMF.MergedOutputType | DMMF.InputType | DMMF.Enum
  isList: boolean
  isRequired: boolean
}
