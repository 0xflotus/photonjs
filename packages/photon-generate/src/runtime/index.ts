export { DMMF } from './dmmf-types'
export { DMMFClass } from './dmmf'
export { deepGet, deepSet } from './utils/deep-set'
export { makeDocument } from './query'
export { Engine } from '@prisma/engine-core'
export { default as debugLib } from 'debug'
