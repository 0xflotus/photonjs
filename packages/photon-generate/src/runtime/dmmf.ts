import { DMMF } from './dmmf-types'
import { keyBy, isScalar, Dictionary, destroyCircular, uniqBy } from './utils/common'

export class DMMFClass implements DMMF.Document {
  datamodel: DMMF.Datamodel
  schema: DMMF.Schema
  mappings: DMMF.Mapping[]
  queryType: DMMF.MergedOutputType
  mutationType: DMMF.MergedOutputType
  outputTypes: DMMF.MergedOutputType[]
  outputTypeMap: Dictionary<DMMF.MergedOutputType> = {}
  inputTypes: DMMF.InputType[]
  inputTypeMap: Dictionary<DMMF.InputType>
  enumMap: Dictionary<DMMF.Enum>
  modelMap: Dictionary<DMMF.Model>
  constructor({ datamodel, schema, mappings }: DMMF.Document) {
    this.datamodel = datamodel
    this.schema = schema
    this.mappings = mappings
    this.enumMap = this.getEnumMap()
    this.queryType = this.getQueryType()
    this.mutationType = this.getMutationType()
    this.schema.outputTypes.push(this.queryType) // create "virtual" query type
    this.schema.outputTypes.push(this.mutationType) // create "virtual" mutation type
    this.modelMap = this.getModelMap()
    this.outputTypes = this.getOutputTypes()

    this.resolveOutputTypes(this.outputTypes)

    this.inputTypes = this.schema.inputTypes
    this.resolveInputTypes(this.inputTypes)
    this.inputTypeMap = this.getInputTypeMap()
    this.resolveFieldArgumentTypes(this.outputTypes, this.inputTypeMap)

    this.outputTypeMap = this.getMergedOutputTypeMap()

    // needed as references are not kept
    this.queryType = this.outputTypeMap['Query']
    this.mutationType = this.outputTypeMap['Mutation']
    this.outputTypes = uniqBy(this.outputTypes, o => o.name).filter(o => {
      return !o.name.endsWith('PreviousValues') && !o.name.includes('Subscription')
    })
    this.inputTypes = uniqBy(this.inputTypes, o => o.name).filter(
      o => !o.name.includes('Subscription') && o.name !== 'MutationType',
    )
  }
  protected outputTypeToMergedOutputType = (outputType: DMMF.OutputType): DMMF.MergedOutputType => {
    const model = this.modelMap[outputType.name]
    return {
      ...outputType,
      isEmbedded: model ? model.isEmbedded : false,
      fields: outputType.fields,
    }
  }
  protected resolveOutputTypes(types: DMMF.MergedOutputType[]) {
    for (const typeA of types) {
      for (const fieldA of typeA.fields) {
        for (const typeB of types) {
          if (typeof fieldA.type === 'string') {
            if (fieldA.type === typeB.name) {
              fieldA.type = typeB
            } else if (this.enumMap[fieldA.type]) {
              fieldA.type = this.enumMap[fieldA.type]
            }
          }
        }
      }
    }
  }
  protected resolveInputTypes(types: DMMF.InputType[]) {
    for (const typeA of types) {
      for (const fieldA of typeA.args) {
        for (const typeB of types) {
          if (typeof fieldA.type === 'string') {
            if (fieldA.type === typeB.name) {
              fieldA.type = typeB
            } else if (this.enumMap[fieldA.type]) {
              fieldA.type = this.enumMap[fieldA.type]
            }
          }
        }
      }
    }
  }
  protected resolveFieldArgumentTypes(types: DMMF.MergedOutputType[], inputTypeMap: Dictionary<DMMF.InputType>) {
    for (const type of types) {
      for (const field of type.fields) {
        for (const arg of field.args) {
          if (typeof arg.type === 'string') {
            if (inputTypeMap[arg.type]) {
              arg.type = inputTypeMap[arg.type]
            } else if (this.enumMap[arg.type]) {
              arg.type = this.enumMap[arg.type]
            }
          }
        }
      }
    }
  }
  protected getQueryType(): DMMF.MergedOutputType {
    return {
      name: 'Query',
      fields: this.schema.queries
        .map(queryToSchemaField)
        .filter(f => !f.name.endsWith('Connection') && f.name !== 'Node'),
      isEmbedded: false,
    }
  }
  protected getMutationType(): DMMF.MergedOutputType {
    return {
      name: 'Mutation',
      fields: this.schema.mutations.map(queryToSchemaField),
      isEmbedded: false,
    }
  }
  protected getOutputTypes(): DMMF.MergedOutputType[] {
    return this.schema.outputTypes.map(this.outputTypeToMergedOutputType)
  }
  protected getEnumMap(): Dictionary<DMMF.Enum> {
    return keyBy(this.schema.enums, e => e.name)
  }
  protected getModelMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.models, m => m.name)
  }
  protected getMergedOutputTypeMap(): Dictionary<DMMF.MergedOutputType> {
    return keyBy(this.outputTypes, t => t.name)
  }
  protected getInputTypeMap(): Dictionary<DMMF.InputType> {
    return keyBy(this.schema.inputTypes, t => t.name)
  }
  public getField(fieldName: string) {
    return (
      // TODO: create lookup table for Query and Mutation
      this.queryType.fields.find(f => f.name === fieldName) || this.mutationType.fields.find(f => f.name === fieldName)
    )
  }
}

const queryToSchemaField = (q: DMMF.Query): DMMF.SchemaField => ({
  name: q.name,
  args: q.args,
  isList: q.output.isList,
  isRequired: q.output.isRequired,
  type: q.output.name,
  kind: 'relation',
})
