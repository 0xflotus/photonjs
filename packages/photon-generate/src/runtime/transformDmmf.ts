import { DMMF } from './dmmf-types'
import { uniqBy, Dictionary, stringifyInputType } from './utils/common'

function transformDmmf(document: DMMF.Document): DMMF.Document {
  filterInputTypes(document)
  filterOutputTypes(document)
  transformInputTypes(document)
  return document
}

function filterInputTypes(document: DMMF.Document) {
  document.schema.inputTypes = uniqBy(document.schema.inputTypes, o => o.name).filter(
    o => !o.name.includes('Subscription') && o.name !== 'MutationType',
  )
}

function filterOutputTypes(document: DMMF.Document) {
  document.schema.outputTypes = uniqBy(document.schema.outputTypes, o => o.name).filter(o => {
    return !o.name.endsWith('PreviousValues') && !o.name.includes('Subscription')
  })
}

function transformInputTypes(document: DMMF.Document) {
  const types = document.schema.inputTypes
  const inputTypes: DMMF.InputType[] = []
  const filterTypes: Dictionary<DMMF.InputType> = {}
  for (const type of types) {
    if (!type.name.endsWith('WhereInput')) {
      inputTypes.push(type)
      continue
    }

    // lastIndexOf necessary if a type is called "WhereInput"
    const index = type.name.lastIndexOf('WhereInput')
    const modelName = type.name.slice(0, index)
    const model = document.datamodel.models.find(m => m.name === modelName)!
    const args = type.args.filter(a => ['AND', 'OR', 'NOT'].includes(a.name))
    // NOTE: list scalar fields don't have where arguments!
    args.push(
      ...model.fields
        .filter(f => !f.isList)
        .map(f => {
          if (!filterTypes[getFilterName(f.type)]) {
            filterTypes[getFilterName(f.type)] = makeFilterType(f.type)
          }
          return {
            name: f.name,
            type: [`${f.type}${f.isList ? 'List' : ''}Filter`],
            isScalar: false,
            isRequired: false,
            isEnum: false,
            isList: false,
          }
        }),
    )
    const newType = {
      name: type.name,
      args,
    }
    console.log(modelName, !!model)
    console.log(stringifyInputType(newType) + '\n')
    inputTypes.push(newType)
  }
  const scalarFilters = Object.values(filterTypes)
  return inputTypes.push(...scalarFilters)
}

function getFilterName(type: string) {
  return `${type}Filter`
}

function makeFilterType(type: string): DMMF.InputType {
  return {
    name: getFilterName(type),
    args: getFilterArgs(type),
  }
}

function getFilterArgs(type: string): DMMF.SchemaArg[] {
  const args: DMMF.SchemaArg[] = getScalarArgs(['equals', 'not'], type)
  switch (type) {
    case 'String':
    case 'ID':
    case 'UUID':
      args.push(...getScalarArgs(['contains', 'startsWith', 'endsWith'], type))
  }

  return args
}

// CONTINUE HERE: Add support for Union Input Types in DMMF, then think about places where we have them:
// - not: {} or not: 5
// - equals: {} or equals: 5
// - id: {} or id: 5

function getBaseFilters(type: string): DMMF.SchemaArg[] {
  return [...getScalarArgs(['equals', 'not'], type), ...getScalarArgs(['AND', 'NOT', 'OR'], getFilterName(type))]
}

function getStringFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['contains', 'startsWith', 'endsWith'], type)
}

function getAlphanumericFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['lt', 'lte', 'gt', 'gte'], type)
}

function getInclusionFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['in', 'notIn'], type)
}

function getScalarArgs(names: string[], type: string): DMMF.SchemaArg[] {
  return names.map(name => getScalarArg(name, type))
}

// {
//   where: {
//     id: {
//       not: {
//         equals: 5
//       }
//     }
//   }
// }

function getScalarArg(name: string, type: string, isList = false): DMMF.SchemaArg {
  return {
    name,
    isEnum: false,
    isList,
    isRequired: false,
    isScalar: true,
    type: [type],
  }
}
