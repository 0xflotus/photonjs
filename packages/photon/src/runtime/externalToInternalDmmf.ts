import { capitalize, lowerCase, plural } from 'prisma-datamodel'
import { DMMF, ExternalDMMF } from './dmmf-types'
import { keyBy } from './utils/common'

function transformFieldKind(model: ExternalDMMF.Model): DMMF.Model {
  return {
    ...model,
    fields: model.fields.map(field => ({ ...field, kind: field.kind === 'relation' ? 'object' : field.kind })),
  }
}

function transformDatamodel(datamodel: ExternalDMMF.Datamodel): DMMF.Datamodel {
  return {
    enums: datamodel.enums,
    models: datamodel.models.map(transformFieldKind),
  }
}

/**
 * Turns type: string into type: string[] for all args in order to support union input types
 * @param document
 */
export function externalToInternalDmmf(document: ExternalDMMF.Document): DMMF.Document {
  return {
    datamodel: transformDatamodel(document.datamodel),
    mappings: getMappings(document),
    schema: transformSchema(document.schema),
  }
}

function transformSchema(schema: ExternalDMMF.Schema): DMMF.Schema {
  return {
    enums: schema.enums,
    inputTypes: schema.inputTypes.map(t => ({ ...t, fields: transformArgs(t.fields) })),
    outputTypes: schema.outputTypes.map(o => ({
      ...o,
      fields: o.fields.map(f => ({ ...f, args: transformArgs(f.args) })),
    })),
  }
}

function transformArgs(args: ExternalDMMF.SchemaArg[]): DMMF.SchemaArg[] {
  return args.map(transformArg)
}

function transformArg(arg: ExternalDMMF.SchemaArg): DMMF.SchemaArg {
  return {
    name: arg.name,
    inputType: [arg.inputType],
  }
}

function getMappings(dmmf: ExternalDMMF.Document): DMMF.Mapping[] {
  return dmmf.datamodel.models.map(model => {
    const modelName = capitalize(model.name)

    const query = dmmf.schema.outputTypes.find(t => t.name === 'Query')
    const mutation = dmmf.schema.outputTypes.find(t => t.name === 'Mutation')

    const queryFields = keyBy(query.fields, f => f.name)
    const mutationFields = keyBy(mutation.fields, f => f.name)

    const mapping: DMMF.Mapping = {
      model: modelName,
    }
    for (const action in DMMF.ModelAction) {
      const queryName = getQueryName(action as DMMF.ModelAction, modelName)
      if (queryFields[queryName] || mutationFields[queryName]) {
        mapping[action] = queryName
      } else {
        mapping[action] = null
      }
    }

    return mapping
  })
}

function getQueryName(action: DMMF.ModelAction, modelName: string) {
  switch (action) {
    case DMMF.ModelAction.findOne:
      return lowerCase(modelName)
    case DMMF.ModelAction.findMany:
      return plural(lowerCase(modelName))
    case DMMF.ModelAction.create:
      return `create${modelName}`
    case DMMF.ModelAction.update:
      return `update${modelName}`
    case DMMF.ModelAction.updateMany:
      return `updateMany${plural(modelName)}`
    case DMMF.ModelAction.upsert:
      return `upsert${modelName}`
    case DMMF.ModelAction.delete:
      return `delete${modelName}`
    case DMMF.ModelAction.deleteMany:
      return `deleteMany${plural(modelName)}`
  }
}
