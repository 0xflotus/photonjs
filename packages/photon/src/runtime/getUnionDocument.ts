import { DMMF } from './dmmf-types'

/**
 * Turns type: string into type: string[] for all args in order to support union input types
 * @param document
 */
export function getUnionDocument(document: DMMF.Document<DMMF.RawSchemaArg>): DMMF.Document<DMMF.SchemaArg> {
  return {
    datamodel: document.datamodel,
    mappings: document.mappings,
    schema: transformSchema(document.schema),
  }
}

function transformSchema(schema: DMMF.Schema<DMMF.RawSchemaArg>): DMMF.Schema<DMMF.SchemaArg> {
  return {
    enums: schema.enums,
    queries: schema.queries.map(q => ({ ...q, args: transformArgs(q.args) })),
    mutations: schema.mutations.map(q => ({ ...q, args: transformArgs(q.args) })),
    inputTypes: schema.inputTypes.map(t => ({ ...t, args: transformArgs(t.args) })),
    outputTypes: schema.outputTypes.map(o => ({
      ...o,
      fields: o.fields.map(f => ({ ...f, type: f.type as string, args: transformArgs(f.args) })),
    })),
  }
}

function transformArgs(args: DMMF.RawSchemaArg[]): DMMF.SchemaArg[] {
  return args.map(transformArg)
}

function transformArg(arg: DMMF.RawSchemaArg): DMMF.SchemaArg {
  return {
    ...arg,
    type: [arg.type],
  }
}
