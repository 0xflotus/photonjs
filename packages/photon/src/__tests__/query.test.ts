import stripAnsi from 'strip-ansi'
import { dmmfDocument } from '../fixtures/example-dmmf'
import { DMMFClass } from '../runtime/dmmf'
import { makeDocument } from '../runtime/query'

describe.skip('validation', () => {
  const dmmf = new DMMFClass(dmmfDocument)
  test('unknown arg, field, incorrect arg type', () => {
    const ast = {
      skip: 200,
      where: {
        name_contains: undefined,
        name_in: ['hans', 'peter', 'schmidt'],
        AND: [
          {
            age_gt: 10123123123,
            this_is_completely_arbitrary: 'veryLongNameGoIntoaNewLineNow@gmail.com',
          },
          {
            age_gt: 10123123123,
            id_endsWith: 'veryLongNameGoIntoaNewLineNow@gmail.com',
            name_contains: 'hans',
            name_gt: 2131203912039123,
            name_in: ['hans'],
            AND: [
              {
                age_gt: '10123123123',
                id_endsWith: 'veryLongNameGoIntoaNewLineNow@gmail.com',
              },
            ],
          },
        ],
      },
      select: {
        id: true,
        name: 'asd',
        name2: true,
        posts: {
          first: 200,
          select: {
            id: true,
            title: false,
          },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'users',
    })
    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, true)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`"Cannot read property 'select' of undefined"`)
    }
  })
  test('missing arg object', () => {
    const ast = {}

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'mutation',
      rootField: 'createPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "

        Invalid \`photon.createPost()\` invocation:

        {
        + data: {
        +   id?: ID,
        +   title: String,
        +   content: String,
        +   author: {
        +     create?: {
        +       id?: ID,
        +       name: String,
        +       strings?: {
        +         set?: String
        +       }
        +     },
        +     connect?: {
        +       id?: ID
        +     }
        +   }
        + }
        }

        Argument data for photon.data is missing. You can see in green what you need to add.

        "
      `)
    }
  })

  test('missing arg object', () => {
    const ast = {}

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'mutation',
      rootField: 'createPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "

        Invalid \`photon.createPost()\` invocation:

        {
        + data: {
        +   id?: ID,
        +   title: String,
        +   content: String,
        +   author: {
        +     create?: {
        +       id?: ID,
        +       name: String,
        +       strings?: {
        +         set?: String
        +       }
        +     },
        +     connect?: {
        +       id?: ID
        +     }
        +   }
        + }
        }

        Argument data for photon.data is missing. You can see in green what you need to add.

        "
      `)
    }
  })

  test('missing arg scalar', () => {
    const ast = {
      data: {
        title: 'string',
        author: {
          connect: { id: '' },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'mutation',
      rootField: 'createPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
                                "

                                Invalid \`photon.createPost()\` invocation:

                                {
                                  data: {
                                    title: 'string',
                                    author: {
                                      connect: {
                                        id: ''
                                      }
                                    },
                                +   content: String,
                                +   id?: ID
                                  }
                                }

                                Argument content for photon.data.content is missing. You can see in green what you need to add.

                                "
                        `)
    }
  })

  test('missing arg scalar && object', () => {
    const ast = {
      data: {
        title: 'string',
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'mutation',
      rootField: 'createPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "

        Invalid \`photon.createPost()\` invocation:

        {
          data: {
            title: 'string',
        +   content: String,
        +   author: {
        +     create?: {
        +       id?: ID,
        +       name: String,
        +       strings?: {
        +         set?: String
        +       }
        +     },
        +     connect?: {
        +       id?: ID
        +     }
        +   },
        +   id?: ID
          }
        }

        Argument content for photon.data.content is missing. You can see in green what you need to add.
        Argument author for photon.data.author is missing. You can see in green what you need to add.

        "
      `)
    }
  })
})
