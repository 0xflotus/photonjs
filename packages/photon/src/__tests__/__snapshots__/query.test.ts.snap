// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation missing arg object 1`] = `
"mutation {
  createPost(data: null) {
    id
    title
    content
  }
}"
`;

exports[`validation missing arg object 3`] = `
"mutation {
  createPost(data: null) {
    id
    title
    content
  }
}"
`;

exports[`validation missing arg scalar && object 1`] = `
"mutation {
  createPost(data: {
    title: \\"string\\"
    content: null
    author: null
    id: null
  }) {
    id
    title
    content
  }
}"
`;

exports[`validation missing arg scalar 1`] = `
"mutation {
  createPost(data: {
    title: \\"string\\"
    author: {
      connect: {
        id: \\"\\"
      }
    }
    content: null
    id: null
  }) {
    id
    title
    content
  }
}"
`;

exports[`validation unknown arg, field, incorrect arg type 1`] = `
"query {
  users(
    skip: 200
    where: {
      name_in: [\\"hans\\", \\"peter\\", \\"schmidt\\"]
      AND: [
        {
          age_gt: 10123123123
          this_is_completely_arbitrary: \\"veryLongNameGoIntoaNewLineNow@gmail.com\\"
        },
        {
          age_gt: 10123123123
          id_endsWith: \\"veryLongNameGoIntoaNewLineNow@gmail.com\\"
          name_contains: \\"hans\\"
          name_gt: 2131203912039123
          name_in: [\\"hans\\"]
          AND: [
            {
              age_gt: \\"10123123123\\"
              id_endsWith: \\"veryLongNameGoIntoaNewLineNow@gmail.com\\"
            }
          ]
        }
      ]
    }
  ) {
    id
    name # INVALID_FIELD
    strings
    name2 # INVALID_FIELD
    posts(first: 200) {
      id
      content
    }
  }
}"
`;
